src/Renderer/Material/Material.cpp

```diff
@@ -32,42 +30,28 @@ namespace SnekVk
 
     Material::Material(Shader* vertexShader, Shader* fragmentShader, u32 shaderCount)
         : vertexShader(vertexShader), fragmentShader(fragmentShader), shaderCount(shaderCount)
-    {
-        auto device = VulkanDevice::GetDeviceInstance();
+    {}
 
-        // TODO: Make this into a separate class 
-        if (descriptorPool == VK_NULL_HANDLE) 
-        {
-            VkDescriptorPoolSize poolSizes[] = {
-                {VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, 10},
-                {VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC, 10},
-                {VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 10},
-                {VK_DESCRIPTOR_TYPE_STORAGE_BUFFER_DYNAMIC, 10}
-            };
+    void Material::SetVertexShader(Shader* shader) 
+    {
+        if (vertexShader == nullptr) shaderCount++;
+        vertexShader = shader; 
+        bufferSize += Buffer::PadUniformBufferSize(shader->GetUniformSize());
 
-            VkDescriptorPoolCreateInfo poolCreateInfo {};
-            poolCreateInfo.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_POOL_CREATE_INFO;
-            poolCreateInfo.flags = 0;
-            poolCreateInfo.maxSets = 10;
-            poolCreateInfo.poolSizeCount = 4;
-            poolCreateInfo.pPoolSizes = poolSizes;
+    }
 
-            SNEK_ASSERT(vkCreateDescriptorPool(device->Device(), &poolCreateInfo, nullptr, &descriptorPool) == VK_SUCCESS,
-                "Unable to create descriptor pool!");
-        }
+    void Material::SetFragmentShader(Shader* shader) 
+    {
+        if (fragmentShader == nullptr) shaderCount++;
+        fragmentShader = shader; 
+        bufferSize += Buffer::PadUniformBufferSize(shader->GetUniformSize());
     }
 
     Material::~Material() 
     {
-        auto device = VulkanDevice::GetDeviceInstance();
+        if (isFreed) return;
```

Comment:
How can we avoid storing this data?

Reply:
Not sure yet. We'll need to really think about how we want our data to be handled since constructors and destructors are a little too simplistic to work in every scenario. We honestly should probably just stick to manual destruction since it gives us more control.